File: clients/android/build-in-docker.pl
#!/usr/bin/perl

use strict;
use File::Path qw(make_path);

die "This script is meant to be run within the camlistore/android Docker contain. Run 'make env' to build it.\n"
    unless $ENV{IN_DOCKER};

my $mode = shift || "debug";

my $ANDROID = "/src/camlistore.org/clients/android";
my $ASSETS = "$ANDROID/assets";
my $GENDIR = "$ANDROID/gen/org/camlistore";

umask 0;
make_path($GENDIR, { mode => 0755 }) unless -d $GENDIR;

$ENV{GOROOT} = "/usr/local/go";
$ENV{GOBIN} = $GENDIR;
$ENV{GOPATH} = "/";
$ENV{GOARCH} = "arm";
print "Building ARM camlistore.org/cmd/camput\n";
system("/usr/local/go/bin/go", "install", "camlistore.org/cmd/camput")
    and die "Failed to build camput";

system("cp", "-p", "$GENDIR/linux_arm/camput", "$ASSETS/camput.arm")
    and die "cp failure";
# TODO: build an x86 version too? if/when those Android devices matter.

{
    open(my $vfh, ">$ASSETS/camput-version.txt") or die "open camput-version error: $!";
    # TODO(bradfitz): make these values automatic, and don't make the
    # "Version" menu say "camput version" when it runs. Also maybe
    # keep a history of these somewhere more convenient.
    print $vfh "app 0.6.1 camput ccacf764 go 70499e5fbe5b";
}

chdir $ASSETS or  die "can't cd to assets dir";

my $digest = `openssl sha1 camput.arm`;
chomp $digest;
print "ARM camput is $digest\n";
die "No digest" unless $digest;
write_file("$GENDIR/ChildProcessConfig.java", "package org.camlistore; public final class ChildProcessConfig { // $digest\n}");

print "Running ant $mode\n";
chdir $ANDROID or die "can't cd to android dir";
exec "ant",
    "-Dsdk.dir=/usr/local/android-sdk-linux",
    "-Dkey.store=/keys/android-camlistore.keystore",
    "-Dkey.alias=camkey",
    $mode;

sub write_file {
    my ($file, $contents) = @_;
    if (open(my $fh, $file)) {
        my $cur = do { local $/; <$fh> };
        return if $cur eq $contents;
    }
    open(my $fh, ">$file") or die "Failed to open $file: $!";
    print $fh $contents;
    close($fh) or die "Close: $!";
    print "Wrote $file\n";
}

File: clients/android/build-in-docker.go
//go:build ignore
// +build ignore

/*
Copyright 2017 The Perkeep Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// This program builds the Perkeep Android application. It is meant to be run
// within the relevant docker container.
package main

import (
	"bufio"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
)

var flagRelease = flag.Bool("release", false, "Whether to assemble the release build, instead of the debug build.")

// TODO(mpl): not sure if the version in app/build.gradle should have anything
// to do with the version we want to use here. look into that later.
const appVersion = "0.7"

var (
	pkDir      = filepath.Join(os.Getenv("GOPATH"), "src/perkeep.org")
	projectDir = filepath.Join(os.Getenv("GOPATH"), "src/perkeep.org/clients/android")
)

func main() {
	flag.Parse()
	if !inDocker() {
		fmt.Fprintf(os.Stderr, "Usage error: this program should be run within a docker container\n")
		os.Exit(2)
	}
	buildPkput()
	buildApp()
}

func buildApp() {
	cmd := exec.Command("./gradlew", "assembleDebug")
	if *flagRelease {
		cmd = exec.Command("./gradlew", "assembleRelease")
	}
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	if _, err := os.ReadFile("./keystore.properties"); err != nil {
		// no keystore.
		// generate one that's in line with the one from devenv/Dockerfile.

		const keystore = `## Code generated by perkeep; DO NOT EDIT.
storeFile=/home/gopher/keystore
storePassword=gopher
keyAlias=perkeep
keyPassword=gopher
`
		err = os.WriteFile("./keystore.properties", []byte(keystore), 0644)
		if err != nil {
			log.Fatalf("could not write default keystore.properties: %+v", err)
		}
		defer os.Remove("./keystore.properties")
	}

	if err := cmd.Run(); err != nil {
		log.Fatalf("Error building Android app: %v", err)
	}
}

func buildPkput() {
	cmd := exec.Command("make")
	cmd.Dir = "./app"
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	err := cmd.Run()
	if err != nil {
		log.Fatalf("could not build pk-put for Android: %+v", err)
	}
}

func version() string {
	return "app " + appVersion + " pk-put " + getVersion() + " " + goVersion()
}

func goVersion() string {
	out, err := exec.Command("go", "version").Output()
	if err != nil {
		log.Fatalf("Error getting Go version with the 'go' command: %v", err)
	}
	return string(out)
}

// getVersion returns the version of Perkeep. Either from a VERSION file at the root,
// or from git.
func getVersion() string {
	slurp, err := ioutil.ReadFile(filepath.Join(pkDir, "VERSION"))
	if err == nil {
		return strings.TrimSpace(string(slurp))
	}
	return gitVersion()
}

var gitVersionRx = regexp.MustCompile(`\b\d\d\d\d-\d\d-\d\d-[0-9a-f]{10,10}\b`)

// gitVersion returns the git version of the git repo at pkDir as a
// string of the form "yyyy-mm-dd-xxxxxxx", with an optional trailing
// '+' if there are any local uncommitted modifications to the tree.
func gitVersion() string {
	cmd := exec.Command("git", "rev-list", "--max-count=1", "--pretty=format:'%ad-%h'",
		"--date=short", "--abbrev=10", "HEAD")
	cmd.Dir = pkDir
	out, err := cmd.Output()
	if err != nil {
		log.Fatalf("Error running git rev-list in %s: %v", pkDir, err)
	}
	v := strings.TrimSpace(string(out))
	if m := gitVersionRx.FindStringSubmatch(v); m != nil {
		v = m[0]
	} else {
		panic("Failed to find git version in " + v)
	}
	cmd = exec.Command("git", "diff", "--exit-code")
	cmd.Dir = pkDir
	if err := cmd.Run(); err != nil {
		v += "+"
	}
	return v
}

func inDocker() bool {
	r, err := os.Open("/proc/self/cgroup")
	if err != nil {
		log.Fatalf(`can't open "/proc/self/cgroup": %v`, err)
	}
	defer r.Close()
	sc := bufio.NewScanner(r)
	for sc.Scan() {
		l := sc.Text()
		fields := strings.SplitN(l, ":", 3)
		if len(fields) != 3 {
			log.Fatal(`unexpected line in "/proc/self/cgroup"`)
		}
		if !(strings.HasPrefix(fields[2], "/docker/") ||
			strings.HasPrefix(fields[2], "/system.slice/docker.service")) {
			return false
		}
	}
	if err := sc.Err(); err != nil {
		log.Fatal(err)
	}
	return true
}
